src/main/scala/is/hail/variant/AltAllele.scala

```diff
@@ -0,0 +1,136 @@
+package is.hail.variant
+
+import is.hail.annotations._
+import is.hail.check._
+import is.hail.expr._
+import is.hail.utils._
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types._
+import org.json4s._
+
+object AltAlleleType extends Enumeration {
+  type AltAlleleType = Value
+  val SNP, MNP, Insertion, Deletion, Complex, Star = Value
+}
+
+object CopyState extends Enumeration {
+  type CopyState = Value
+  val Auto, HemiX, HemiY = Value
+}
+
+object AltAllele {
+
+  def fromRegionValue(m: MemoryBuffer, offset: Long): AltAllele = {
+    val t = TAltAllele.representation()
+    val ref = TString.loadString(m, t.loadField(m, offset, 0))
+    val alt = TString.loadString(m, t.loadField(m, offset, 1))
+    AltAllele(ref, alt)
+  }
+
+  def sparkSchema: StructType = StructType(Array(
+    StructField("ref", StringType, nullable = false),
+    StructField("alt", StringType, nullable = false)))
+
+  def fromRow(r: Row): AltAllele =
+    AltAllele(r.getString(0), r.getString(1))
+
+  def gen(ref: String): Gen[AltAllele] =
+    for (alt <- Gen.frequency((10, genDNAString),
+      (1, Gen.const("*"))) if alt != ref)
+      yield AltAllele(ref, alt)
+
+  def gen: Gen[AltAllele] =
+    for (ref <- genDNAString;
+      alt <- genDNAString if alt != ref)
+      yield AltAllele(ref, alt)
+
+  implicit def altAlleleOrder: Ordering[AltAllele] = new Ordering[AltAllele] {
+    def compare(x: AltAllele, y: AltAllele): Int = x.compare(y)
+  }
+}
+
+trait IAltAllele {
```

Comment:
This is the key change, I added an `IAltAllele` with all the meaningful methods. Then `AltAllele` is just a data carrier inheriting all the relevant methods.